### PR TITLE
xshards-pandas.md

### PR DESCRIPTION
**Pull request by Le-Zheng**

### CODE REVIEW

**Comment by Le-Zheng** [comment]
pre-review https://le-test-xshards.readthedocs.io/en/latest/doc/UseCase/xshards-pandas.html

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 1]
```diff
@@ -2,4 +2,124 @@
 
 ---
 
-**In this guide we will describe how to use XShards to scale-out Pandas data processing for distribtued deep learning.**
\ No newline at end of file
+![](../../../image/colab_logo_32px.png)[Run in Google Colab](https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_xshards_pandas.ipynb) &nbsp;![](../../../image/GitHub-Mark-32px.png)[View source on GitHub](https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ncf_xshards_pandas.ipynb)
+
+---
+
+**In this guide we will describe how to use XShards to scale-out Pandas data processing for distribtued deep learning.**
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:9]: Add a link to `Orca/Overview/data-parallel-processing.md`

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 2]
```diff
@@ -2,4 +2,124 @@
...
+
+We will introduce how to use `XShards` and *existing* python code to process [movielens-1M](https://grouplens.org/datasets/movielens/1m/) in a distributed and data-parallel fashion.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:11]: remove this sentence

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 3]
```diff
@@ -2,4 +2,124 @@
...
+
+---
+
+**In this guide we will describe how to use XShards to scale-out Pandas data processing for distribtued deep learning.**
+
+We will introduce how to use `XShards` and *existing* python code to process [movielens-1M](https://grouplens.org/datasets/movielens/1m/) in a distributed and data-parallel fashion. 
+
+### **1. Read input data into XShards of Pandas DataFrame**
+
+First, read CVS, JSON or Parquet files (stored on local disk, HDFS, AWS S3, etc.) to obtain an `XShards` of Pandas Dataframe, as shown below:
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:15]: First, read CVS, JSON or Parquet files to to obtain an `XShards` of Pandas Dataframe (i.e., an distributed and sharded dataset where each partition contained a Pandas Dataframe), as shown below:

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 4]
```diff
@@ -2,4 +2,124 @@
...
+```python
+from zoo.orca.data.pandas import read_csv
+full_data = read_csv(new_rating_files, sep=':', header=None,
+                     names=['user', 'item', 'label'], usecols=[0, 1, 2],
+                     dtype={0: np.int32, 1: np.int32, 2: np.int32})
+```
+
+### **2. Process Pandas Dataframes using XShards**
+
+Next, use XShards to efficiently process large-size Pandas Dataframes in a distributed and data-parallel fashion. Run Python codes on each partition in a data-parallel fashion using `XShards.transform_shard`, as shown below:
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:26]: You may run standard Python code on each partition in a data-parallel fashion using `XShards.transform_shard`, as shown below:

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 5]
```diff
@@ -2,4 +2,124 @@
...
+def split_train_test(data):
+  train, test = train_test_split(data, test_size=0.2, random_state=100)
+  return train, test
+
+train_data, test_data = full_data.transform_shard(split_train_test).split()
+```
+
+### **3. Define NCF model**
+
+This example defines NCF model in a TensorFlow 1.15 program.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:50]: Define the NCF model using TensorFlow 1.15 APIs:

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 6]
```diff
@@ -2,4 +2,124 @@
...
+        with tf.name_scope("optimzation"):
+            self.optim = tf.train.AdamOptimizer(1e-3, name='Adam')
+            self.optimizer = self.optim.minimize(self.loss)
+
+embedding_size=16
+model = NCF(embedding_size, max_user_id, max_item_id)
+```
+### **4. Fit with Orca Estimator**
+
+Finally, run distributed model training/inference on the XShards of Pandas DataFrames directly.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:103]: Finally, directly run distributed model training/inference on the XShards of Pandas DataFrames.

File: docs/readthedocs/source/doc/UseCase/xshards-pandas.md [hunk 7]
```diff
@@ -6,13 +6,11 @@
...
 
-**In this guide we will describe how to use XShards to scale-out Pandas data processing for distribtued deep learning.**
-
-We will introduce how to use `XShards` and *existing* python code to process [movielens-1M](https://grouplens.org/datasets/movielens/1m/) in a distributed and data-parallel fashion. 
+**In this guide we will describe how to use [XShards](../Orca/Overview/data-parallel-processing.md) to scale-out Pandas data processing for distribtued deep learning.** 
 
 ### **1. Read input data into XShards of Pandas DataFrame**
 
-First, read CVS, JSON or Parquet files (stored on local disk, HDFS, AWS S3, etc.) to obtain an `XShards` of Pandas Dataframe, as shown below:
+First, read CVS, JSON or Parquet files to to obtain an `XShards` of Pandas Dataframe (i.e., an distributed and sharded dataset where each partition contained a Pandas Dataframe), as shown below:
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/UseCase/xshards-pandas.md:13]: First, read CVS, JSON or Parquet files into an `XShards` of Pandas Dataframe (i.e., an distributed and sharded dataset where each partition contains a Pandas Dataframe),

**Comment by jason-dai** [comment]
See comments above; others LGTM.

File: docs/docs/colab-notebook/orca/quickstart/ncf_xshards_pandas.ipynb [hunk 8]
```diff
@@ -334,77 +334,50 @@
         "id": "MtmSkXcNS4jy"
       },
       "source": [
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ncf_xshards_pandas.ipynb:336]: there is an empty cell with `stderr` above the cell of `Define NCF Model`

**Comment by jason-dai** [comment]
LGTM

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Quickstart-notebook/323/
